### DISCUSSION TITLE
Initialization Vector in AES

### DISCUSSION THREAD

**Discussion by ThomasETN** [whyoleg/cryptography-kotlin#9] [Q&A] [answered]
Hello, may I ask how we could specify the initialization vector to be used, during AES-CBC decryption?
I searched for "iv" "initialization vector" etc. in the API docs but I couldn't find anything.

Thanks!

**Comment by whyoleg** [comment, marked as answer]
Hey! Thanks for the question, looks like I need to update documentation around this.
As for now, the only way for this is to prefix the encrypted message with IV. So you should do something like 'cipher.decrypt(iv+ciphertext)'.
I still not sure what is the best way to provide better symmetric (between encryption and decryption) API for this, so it will be safe enough and easy to use

**Reply by ThomasETN** [reply]
Thanks it worked, but what about encryption? I can't specify the iv with the same format:
 `cipher.encrypt(initializationVector + plainText)` returns a different result every time 🤔

**Reply by ThomasETN** [reply]
The IV is an essential part of block cipher modes in AES, why not make it an input parameter for all the functions?

**Reply by whyoleg** [reply]
Yes, as for now, there is no way to provide custom IV during encryption - IV will be generated automatically via `SecureRandom` and prefixed to encrypted data, so: `cipher.encrypt(plaintext) = iv + ciphertext`
While I understand, that IV is an essential part, still, if IV will be a parameter, it will be easy to miss-use API and reuse the same IV everywhere. Im not sure, what is the best way here, on how to provide better API for such use cases.

**Reply by ThomasETN** [reply]
In my particular use case I am working on a legacy system and I have to reuse the same IV.

(the system requires 128 bits of zero as IV 🥴)
(actually the system requires AES-ECB and I am attempting to simulate ECB with CBC by chunking the plaintext into single blocks, making it "effectively ECB" 🥴🥴)

**Reply by ThomasETN** [reply]
I have moved on to using the good ol' `expect`/`actual` pair in my project.

Thanks for the support once again!

**Reply by whyoleg** [reply]
Yeah, I understand the inconvenience. I would like to improve here, but wanted to do it somehow clear, that this could be unsafe and such an API should be used very carefully.

Thanks for question, and hopefully someday even this use-case will be supported!
